rir/src/compiler/translations/pir_2_rir.cpp

```diff
@@ -26,73 +28,499 @@ namespace pir {
+                    todo.insert(pre);
+                } else {
+                    merge(pre, accumulated);
+                    mergePhiInp(pre);
+                }
+            }
+        }
+
+        if (verbose) {
+            for (auto ll : livenessInterval) {
+                auto& l = ll.second;
+                ll.first->printRef(std::cout);
+                std::cout << " is live : ";
+                for (size_t i = 0; i < bbsSize; ++i) {
+                    if (l[i].live) {
+                        std::cout << i << " [";
+                        std::cout << l[i].begin << ",";
+                        std::cout << l[i].end << "]  ";
+                    }
+                }
+                std::cout << "\n";
+            }
+        }
+    }
+
+    void computeStackAllocation() {
+        Visitor::run(code->entry, [&](BB* bb) {
+            {
+                // Phi at the beginning of BB, all inputs at the end of
+                // immediate predecessors -> put on stack
+                size_t pos = 1;
+                for (auto i : *bb) {
+                    Phi* phi = Phi::Cast(i);
+                    if (!phi)
+                        break;
+                    bool argsInRightOrder = true;
+                    phi->eachArg([&](BB* in, Value* v) {
+                        if (in->next0 != bb || in->size() < pos ||
+                            *(in->end() - pos) != v) {
+                            argsInRightOrder = false;
+                        }
+                    });
+                    if (!argsInRightOrder)
+                        break;
+                    phi->eachArg(
+                        [&](BB*, Value* v) { allocation[v] = stackSlot; });
+                    allocation[phi] = stackSlot;
+                    pos++;
+                }
+            }
+
+            // Precolor easy stack load-stores within one BB
+            size_t pos = 0;
+            std::deque<Instruction*> stack;
+
+            auto tryLoadingArgsFromStack = [&](Instruction* i) {
+                if (i->nargs() == 0 || stack.size() < i->nargs())
+                    return;
+
+                // Match all args to stack slots.
+                size_t newStackSize = stack.size();
+                bool foundAll = true;
+                auto check = stack.rbegin();
+                i->eachArg([&](Value* arg) {
+                    while (check != stack.rend() && *check != arg) {
+                        ++check;
+                        --newStackSize;
+                    }
+
+                    if (check == stack.rend()) {
+                        foundAll = false;
+                    } else {
+                        // found arg!
+                        ++check;
+                        --newStackSize;
+                    }
                 });
+
+                if (!foundAll)
+                    return;
+
+                // pop args from stack, discarding all unmatched values
+                // in the process
```

Comment:
I don't understand this, how can we discard the values that are not matched?
I mean, if the stack contains `[... A B C]` and the instruction expects `A` and `C`, I'd think we cannot allocate `A` to stack, but this looks like we do and we throw away `B`

Reply:
what we do here is: the stack contains `xxx A B C` and they are all only used once. then we resize the stack to `xxx`, mark `A, C` to be on the stack and put `B` in a register

Reply:
Oh right, this is not what's actually on the stack... The B will get stored to a local by its producer in this case. Thanks for the explanation!